storage/buffer.go

```diff
@@ -199,26 +194,97 @@ func (b *dbBuffer) ReadEncoded(ctx context.Context, start, end time.Time) [][]xi
 		}
 
 		unmerged := make([]xio.SegmentReader, 0, len(bucket.encoders))
-		for i := range bucket.encoders {
-			stream := bucket.encoders[i].encoder.Stream()
-			if stream == nil {
-				// TODO(r): log an error and emit a metric, this is pretty bad as this
-				// encoder should have values if "shouldRead" returned true
-				continue
+		b.readBucketStreams(ctx, bucket, now, current, func(stream xio.SegmentReader) {
+			unmerged = append(unmerged, stream)
+		})
+
+		results = append(results, unmerged)
```

Comment:
Is unmerged guarenteed to be of some size?  Probably need to avoid adding nil entries to results here.